timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,480 @@
+    std::vector<int> v(s.begin(), s.end());
+    return v;
+}
+
+template <typename RealType, bool Interpolated>
+NonbondedInteractionGroup<RealType, Interpolated>::NonbondedInteractionGroup(
+    const std::set<int> &row_atom_idxs,
+    const std::vector<int> &lambda_plane_idxs,  // [N]
+    const std::vector<int> &lambda_offset_idxs, // [N]
+    const double beta,
+    const double cutoff,
+    const std::string &kernel_src)
+    : N_(lambda_offset_idxs.size()), NR_(row_atom_idxs.size()), NC_(N_ - NR_), cutoff_(cutoff), nblist_(NC_, NR_),
+      beta_(beta), d_sort_storage_(nullptr), d_sort_storage_bytes_(0), nblist_padding_(0.1), disable_hilbert_(false),
+      kernel_ptrs_({// enumerate over every possible kernel combination
+                    // U: Compute U
+                    // X: Compute DU_DL
+                    // L: Compute DU_DX
+                    // P: Compute DU_DP
+                    //                             U  X  L  P
+                    &k_nonbonded_unified<RealType, 0, 0, 0, 0>,
+                    &k_nonbonded_unified<RealType, 0, 0, 0, 1>,
+                    &k_nonbonded_unified<RealType, 0, 0, 1, 0>,
+                    &k_nonbonded_unified<RealType, 0, 0, 1, 1>,
+                    &k_nonbonded_unified<RealType, 0, 1, 0, 0>,
+                    &k_nonbonded_unified<RealType, 0, 1, 0, 1>,
+                    &k_nonbonded_unified<RealType, 0, 1, 1, 0>,
+                    &k_nonbonded_unified<RealType, 0, 1, 1, 1>,
+                    &k_nonbonded_unified<RealType, 1, 0, 0, 0>,
+                    &k_nonbonded_unified<RealType, 1, 0, 0, 1>,
+                    &k_nonbonded_unified<RealType, 1, 0, 1, 0>,
+                    &k_nonbonded_unified<RealType, 1, 0, 1, 1>,
+                    &k_nonbonded_unified<RealType, 1, 1, 0, 0>,
+                    &k_nonbonded_unified<RealType, 1, 1, 0, 1>,
+                    &k_nonbonded_unified<RealType, 1, 1, 1, 0>,
+                    &k_nonbonded_unified<RealType, 1, 1, 1, 1>}),
+      compute_w_coords_instance_(kernel_cache_.program(kernel_src.c_str()).kernel("k_compute_w_coords").instantiate()),
+      compute_permute_interpolated_(
+          kernel_cache_.program(kernel_src.c_str()).kernel("k_permute_interpolated").instantiate()),
+      compute_add_du_dp_interpolated_(
+          kernel_cache_.program(kernel_src.c_str()).kernel("k_add_du_dp_interpolated").instantiate()) {
+
+    if (NR_ == 0) {
+        throw std::runtime_error("row_atom_idxs must be nonempty");
+    }
+
+    // compute set of column atoms as set difference
+    std::vector<int> all_atom_idxs(N_);
+    std::iota(all_atom_idxs.begin(), all_atom_idxs.end(), 0);
+    std::set<int> col_atom_idxs;
+    std::set_difference(
+        all_atom_idxs.begin(),
+        all_atom_idxs.end(),
+        row_atom_idxs.begin(),
+        row_atom_idxs.end(),
+        std::inserter(col_atom_idxs, col_atom_idxs.end()));
+
+    if (lambda_offset_idxs.size() != N_) {
```

Comment:
nit: Move lambda offset and plane checks before construction of column atoms

Reply:
good call, moved in efa4828